### ISSUE TITLE
Support for pTeX

### ISSUE THREAD

**Issue by josephwright** [latex3/latex3#230]
Currently we cover pdfTeX/XeTeX/LuaTeX only

**Comment by josephwright** [comment]
First issue: `uptex` runs upTeX while `uplatex` is a format build using e-upTeX. This means `l3build` goes wrong when trying to select the correct format (`uptex -fmt=uplatex` mismatches engine and format files).

**Comment by josephwright** [comment]
Perhaps simply special-case that into `l3build`?

**Comment by josephwright** [comment]
Forcing the issue with a bit of a hack, we can get to a list of failures with `expl3`:
```
  - ../build/test/d3dvipdfmx.uplatex.diff
  - ../build/test/d3dvips.uplatex.diff
  - ../build/test/d3pdfmode.uplatex.diff
  - ../build/test/d3xdvipdfmx.uplatex.diff
  - ../build/test/m3box001.uplatex.diff
  - ../build/test/m3box002.uplatex.diff
  - ../build/test/m3box003.uplatex.diff
  - ../build/test/m3box004.uplatex.diff
  - ../build/test/m3coffins001.uplatex.diff
  - ../build/test/m3expl001.uplatex.diff
  - ../build/test/m3expl002.uplatex.diff
  - ../build/test/m3expl003.uplatex.diff
  - ../build/test/m3expl004.uplatex.diff
  - ../build/test/m3file001.uplatex.diff
  - ../build/test/m3int001.uplatex.diff
  - ../build/test/m3int002.uplatex.diff
  - ../build/test/m3keys002.uplatex.diff
  - ../build/test/m3skip001.uplatex.diff
  - ../build/test/m3skip003.uplatex.diff
  - ../build/test/m3skip005.uplatex.diff
  - ../build/test/m3tl013.uplatex.diff
```
These break down into:
 - Register number differences (handle with an engine-selective register gobble?)
 - Direction information `, yoko direction` (normalise out?)
 - Driver differences (probably needs checking in `l3drivers` then treating as engine-dependent)
 - Some file loading by the format (`upldefs.ltx` and various `.fd` files)
 - Some weird issues with `m3tl013`: some chars are dropped or mishandled, may be engine issues

**Comment by josephwright** [comment]
Presumably we will want to work out where this fits in to https://github.com/latex3/svn-mirror/issues/233

**Comment by kuroky49** [comment]
David said as follows in private communication:
> Sadly none of us read Japanese so currently the only way to see what primitives are available in
the ptex variants is to read the web .ch source files. So a simple list of the extra primitives available in each variant would be a very useful thing to have.

Would be helpful the webpage http://ascii.asciimw.jp/pb/ptex/help/zz.html?  Machine translation (e.g., Google Translation) will help you to understand what it says.  If you need some help, I could translate some contents.

**Comment by josephwright** [comment]
Thanks for the link: at present we only really need the names of the primitives so that should be fine. (I did a sweep over the `.ch` files but will double-check against the web page.)

**Comment by davidcarlisle** [comment]
@kuroky49 thanks that looks very useful! 

google translate may not be perfect but it seems to make a perfectly understandable version of that page.

**Comment by h-kitagawa** [comment]
Hi,
> Some weird issues with m3tl013: some chars are dropped or mishandled, may be engine issues

I tried to solve this issue (https://github.com/h-kitagawa/svn-mirror/commit/a3a0ba11f6b0ab0201adfef8715015d04ecd9737).
 * upTeX's "non-CJK part" is not Unicode-aware.
 * By default, greek letters such as Σ are treated as Japanese characters, 
   so I added `\uptex_disablecjktoken:D` to the test for preventing it.

**Comment by josephwright** [comment]
@h-kitagawa Thanks for that: I'm working out various bits and pieces which go together (testing, conditionals, drivers, ...) and once we know what we want I'll update `expl3`. Should not take very long: probably by the end of the week.

**Comment by josephwright** [comment]
Fixed by fee8c77fcd5c59dd5d04c69416577d079bd9ddab.
